test/fail_compilation/ice17831.d

```diff
@@ -0,0 +1,69 @@
+// REQUIRED_ARGS: -d
+/*
+TEST_OUTPUT:
+---
+fail_compilation/ice17831.d(22): Error: case variables cannot be declared in switch bodies
+fail_compilation/ice17831.d(20):        variable for case declared here
+fail_compilation/ice17831.d(36): Error: case variables cannot be declared in switch bodies
+fail_compilation/ice17831.d(34):        variable for case declared here
+fail_compilation/ice17831.d(51): Error: case variables cannot be declared in switch bodies
+fail_compilation/ice17831.d(48):        variable for case declared here
+fail_compilation/ice17831.d(64): Error: case variables cannot be declared in switch bodies
+fail_compilation/ice17831.d(63):        variable for case declared here
+---
+ */
+
+void test17831a()
+{
+    switch (0)
+    {
+        foreach (i; 0 .. 5)
+        {
+            case i:
+                break;
+        }
+        default:
+            break;
+    }
+}
+
+void test17831b()
+{
+    switch (0)
+    {
+        for (int i = 0; i < 5; i++)
+        {
+            case i:
+                break;
+        }
+        default:
+            break;
+    }
+}
+
+void test17831c()
+{
+    switch (0)
+    {
+        int i = 0;
+        while (i++ < 5)
+        {
+            case i:
```

Comment:
@WalterBright - thinking a bit more about this. Maybe it would be better to explicitly remove the switch statement reference from `Scope` when pushing a new for loop bodies (`sc = sc.push(); sc.sw = null;`), thus disallowing cases outright from appearing inside loop bodies.

This would be a bit more breaking, but I can't think of any justification for allowing this.